drivers/arm/serial_usart.c

```diff
@@ -0,0 +1,198 @@
+#elif SELECT_SOFT_SERIAL_SPEED == 3
+#    define USART_SPEED 38400
+#elif SELECT_SOFT_SERIAL_SPEED == 4
+#    define USART_SPEED 19200
+#elif SELECT_SOFT_SERIAL_SPEED == 5
+#    define USART_SPEED 14400
+#else
+#    error invalid SELECT_SOFT_SERIAL_SPEED value
+#endif
+
+#define TIMEOUT 100
+
+static inline void sdClear(SerialDriver* driver) {
+    while (sdGetTimeout(driver, TIME_IMMEDIATE) != MSG_TIMEOUT) {
+        // Do nothing with the data
+    }
+}
+
+static inline msg_t sdPutHalfDuplex(SerialDriver* driver, uint8_t data) {
+    msg_t ret = sdPut(driver, data);
+    sdGet(driver);
+
+    return ret;
+}
+
+static inline msg_t sdWriteTimeoutHalfDuplex(SerialDriver* driver, uint8_t* data, uint8_t size, uint32_t timeout) {
+    msg_t ret = sdWriteTimeout(driver, data, size, timeout);
+
+    // Half duplex requires us to read back the data we just wrote - just throw it away
+    uint8_t dump[size];
+    sdRead(driver, dump, size);
+
+    return ret;
+}
+
+static inline msg_t sdWriteHalfDuplex(SerialDriver* driver, uint8_t* data, uint8_t size) {
+    msg_t ret = sdWrite(driver, data, size);
+
+    // Half duplex requires us to read back the data we just wrote - just throw it away
+    uint8_t dump[size];
+    sdRead(driver, dump, size);
+
+    return ret;
+}
+
+static const SerialConfig sdcfg = {
+    USART_SPEED,     // speed - mandatory
+    0,               // CR1
+    0,               // CR2
+    USART_CR3_HDSEL  // CT3
+};
+
+void handle_soft_serial_slave(void);
+
+/*
+ * This thread runs on the slave and responds to transactions initiated
+ * by the master
+ */
+static THD_WORKING_AREA(waSlaveThread, 2048);
+static THD_FUNCTION(SlaveThread, arg) {
+    (void)arg;
+    chRegSetThreadName("slave_transport");
+
+    while (true) {
+        handle_soft_serial_slave();
+    }
+}
+
+void usart_init(void) {
+    sdStart(&USART_DRIVER, &sdcfg);
+
+#if defined(USE_GPIOV1)
+    palSetPadMode(PAL_PORT(SOFT_SERIAL_PIN), PAL_PAD(SOFT_SERIAL_PIN), PAL_MODE_STM32_ALTERNATE_OPENDRAIN);
+#else
+    palSetPadMode(PAL_PORT(SOFT_SERIAL_PIN), PAL_PAD(SOFT_SERIAL_PIN), PAL_MODE_ALTERNATE(USART_TX_PAL_MODE) | PAL_STM32_OTYPE_OPENDRAIN);
+#endif
+}
+
+void usart_slave_init(void) {
+    usart_init();
+
+    // Start transport thread
+    chThdCreateStatic(waSlaveThread, sizeof(waSlaveThread), HIGHPRIO, SlaveThread, NULL);
+}
+
+static SSTD_t* Transaction_table      = NULL;
+static uint8_t Transaction_table_size = 0;
+
+void soft_serial_initiator_init(SSTD_t* sstd_table, int sstd_table_size) {
+    Transaction_table      = sstd_table;
+    Transaction_table_size = (uint8_t)sstd_table_size;
+
+    usart_init();
+}
+
+void soft_serial_target_init(SSTD_t* sstd_table, int sstd_table_size) {
+    Transaction_table      = sstd_table;
+    Transaction_table_size = (uint8_t)sstd_table_size;
+
+    usart_slave_init();
+}
+
+void handle_soft_serial_slave(void) {
+    uint8_t sstd_index = sdGet(&USART_DRIVER);  // first chunk is always transaction id
```

Comment:
In theory, this code shouldn't be called outside of `transport.c` so *should* be main thread. I can definitely add the locking, but i cant see a use case for it right now? I2C could try and talk to multiple devices, but if that were to happen here it would lead to bad times.